rust/arrow/src/array/array.rs

```diff
@@ -1647,6 +1671,221 @@ impl From<(Vec<(Field, ArrayRef)>, Buffer, usize)> for StructArray {
+            unsafe { Some(Some((&*self.ptr.offset((i + n) as isize)).clone())) }
+        }
+    }
+}
+
+impl<'a, K: ArrowPrimitiveType> DictionaryArray<K> {
+    /// Return an iterator to the keys of this dictionary.
+    pub fn keys(&'a self) -> NullableIter<'a, K::Native> {
+        NullableIter::<'a, K::Native> {
+            data: &self.data,
+            ptr: unsafe { self.raw_values.get().offset(self.data.offset() as isize) },
+            i: 0,
+            len: self.data.len(),
+        }
+    }
+
+    /// Returns an `ArrayRef` to the dictionary values.
+    pub fn values(&self) -> ArrayRef {
+        self.values.clone()
+    }
+
+    /// Returns a clone of the value type of this list.
+    pub fn value_type(&self) -> DataType {
+        self.values.data().data_type().clone()
+    }
+
+    /// The length of the dictionary is the length of the keys array.
+    pub fn len(&self) -> usize {
+        self.data.len()
+    }
+
+    pub fn is_ordered(&self) -> bool {
+        self.is_ordered
+    }
+}
+
+/// Constructs a `DictionaryArray` from an array data reference.
+impl<T: ArrowPrimitiveType> From<ArrayDataRef> for DictionaryArray<T> {
+    fn from(data: ArrayDataRef) -> Self {
+        assert_eq!(
+            data.buffers().len(),
+            1,
+            "DictionaryArray data should contain a single buffer only (keys)."
+        );
+        assert_eq!(
+            data.child_data().len(),
+            1,
+            "DictionaryArray should contain a single child array (values)."
+        );
+
+        let raw_values = data.buffers()[0].raw_data();
+        let dtype: &DataType = data.data_type();
+        let values = make_array(data.child_data()[0].clone());
+        if let DataType::Dictionary(_, _) = dtype {
+            Self {
+                data: data,
+                raw_values: RawPtrBox::new(raw_values as *const T::Native),
+                values: values,
+                is_ordered: false,
+            }
+        } else {
+            panic!("DictionaryArray must have Dictionary data type.")
+        }
+    }
+}
+
+/// Constructs a `DictionaryArray` from an iterator of optional strings.
+impl<T: ArrowPrimitiveType> FromIterator<Option<&'static str>> for DictionaryArray<T> {
+    fn from_iter<I: IntoIterator<Item = Option<&'static str>>>(iter: I) -> Self {
+        let iter = iter.into_iter();
+        let (lower, _) = iter.size_hint();
+        let key_builder = PrimitiveBuilder::<T>::new(lower);
+        let value_builder = StringBuilder::new(256);
+        // Note: "true" here reserves one value element for null.
+        let mut builder = StringDictionaryBuilder::new(key_builder, value_builder, true);
+        for i in iter {
+            if let Some(i) = i {
+                // Note: impl ... for Result<DictionaryArray<T>> fails with
+                // error[E0117]: only traits defined in the current crate can be implemented for arbitrary types
+                builder.append(i).unwrap();
```

Comment:
Thanks for taking the time to document issue here, maybe let's use `expect` instead of `unwrap` though?

Reply:
I'll weed out the unwraps pro-temp, but proper error handling would be better.

I need to inspect the `Vec` `FromIterator` implementation to see how they work around the issue raised.